src/webgpu/shader/validation/expression/call/builtin/distance.spec.ts

```diff
@@ -0,0 +1,149 @@
+const builtin = 'distance';
+export const description = `
+Validation tests for the ${builtin}() builtin.
+`;
+
+import { makeTestGroup } from '../../../../../../common/framework/test_group.js';
+import { keysOf, objectsToRecord } from '../../../../../../common/util/data_tables.js';
+import {
+  Type,
+  kConvertableToFloatScalarsAndVectors,
+  scalarTypeOf,
+  ScalarType,
+} from '../../../../../util/conversion.js';
+import { QuantizeFunc, quantizeToF16, quantizeToF32 } from '../../../../../util/math.js';
+import { ShaderValidationTest } from '../../../shader_validation_test.js';
+
+import {
+  fullRangeForType,
+  kConstantAndOverrideStages,
+  stageSupportsType,
+  validateConstOrOverrideBuiltinEval,
+} from './const_override_validation.js';
+
+export const g = makeTestGroup(ShaderValidationTest);
+
+const kValidArgumentTypes = objectsToRecord(kConvertableToFloatScalarsAndVectors);
+
+function quantizeFunctionForScalarType(type: ScalarType): QuantizeFunc<number> {
+  switch (type) {
+    case Type.f32:
+      return quantizeToF32;
+    case Type.f16:
+      return quantizeToF16;
+    default:
+      return (v: number) => v;
+  }
+}
+
+g.test('values')
+  .desc(
+    `
+Validates that constant evaluation and override evaluation of ${builtin}() never errors
+`
+  )
+  .params(u =>
+    u
+      .combine('stage', kConstantAndOverrideStages)
+      .combine('type', keysOf(kValidArgumentTypes))
+      .filter(u => stageSupportsType(u.stage, kValidArgumentTypes[u.type]))
+      .beginSubcases()
+      .expand('a', u => fullRangeForType(kValidArgumentTypes[u.type], 5))
+      .expand('b', u => fullRangeForType(kValidArgumentTypes[u.type], 5))
+  )
+  .beforeAllSubcases(t => {
+    if (scalarTypeOf(kValidArgumentTypes[t.params.type]) === Type.f16) {
+      t.selectDeviceOrSkipTestCase('shader-f16');
+    }
+  })
+  .fn(t => {
+    let expectedResult = true;
+
+    const scalarType = scalarTypeOf(kValidArgumentTypes[t.params.type]);
+    const quantizeFn = quantizeFunctionForScalarType(scalarType);
+
+    // Distance equation: length(a - b)
+    // Should be invalid if the calculations result in intermediate values that
+    // exceed the maximum representable float value for the given type.
+    const a = Number(t.params.a);
+    const b = Number(t.params.b);
+    const ab = quantizeFn(a - b);
+
+    if (!Number.isFinite(ab)) {
+      expectedResult = false;
+    }
+
+    // Only calculates the full length if the type is a vector. Otherwise abs(a-b) is used.
+    if (kValidArgumentTypes[t.params.type].width > 1) {
+      const ab2 = quantizeFn(ab * ab);
+      const sqrLen = quantizeFn(ab2 * kValidArgumentTypes[t.params.type].width);
+      // Square root does not need to be calculated because it can never fail if
+      // the previous results are finite.
+
+      if (!Number.isFinite(ab2) || !Number.isFinite(sqrLen)) {
```

Comment:
What about DBZ when the length is zero?

Reply:
Given that this isn't `normalize`, the input vector doesn't need to be divided by the length, the result is the length itself. And square root is fine because `sqrt(0) == 0`. Unless I'm missing something?

Reply:
Sorry, getting my builtins mixed up. Ignore me.